Fast lexer

Kronos3 opened the pull request:
This PR reimplements the lexing strategy that produces a hugely faster lexer (8-10x speedup). It uses the same design pattern as can be found in the Scala compiler using a manually written recursive descent parser that moves through the file character-by-character. I also included some fixes in the parser that speed up the parsing by ~30% using IntelliJ's static analysis and auto-fixes. Mainly the fixes removed some redunant syntax and added `private` to functions that could be internalized (allows inlining and other optimizations by Scala).

## Notable changes

There are some very minor discrepancies between the old and new lexer, @bocchino I didn't get exact feature parity because some behavior seems a little under defined so I'll wait for your input:

1. Trailing spaces after a `\` line continuation are disallowed. There is a single file in the FPrime repo that has some trailing spaces which throws an error in the lexer. I'm leaning toward keeping the functionality as is since it enforces cleaner syntax.
2. The lexer uses an error `Context` to track invalid tokens or tokenization errors.
3.  The string trimming of multi-line strings (3-strings) is pretty confusing (or bugged). The old behavior had odd behavior:

```
constant F: string = """
    indented
    keeps
    trailing newline
    """
```

Generates a string with value: `indented\nkeeps\ntrailing newline\n`

```
constant F: string = """
non-indented
does not keep
trailing newline
```

Generates string with value: `non-indented\ndoes not keep\ntrailing newline`

For now the new lexer will not string trailing newlines ever but I can easily add that in.

## Benchmarks

To benchmark I ran a JVM-based profiler to compare lexing times on the locs generation step of the FppTest build (this is a parsing intensive build stage). The images screenshots of the runtime of the lexer before and after this PR. I also `time`ed the runtime of the native GraalVM build of `fpp-locate-defs` before and after this PR. The stats are included below.

Before:
![image](https://github.com/user-attachments/assets/61bc0f98-fbfe-46ff-8581-2e9e2573119d)

```
real	0m3.460s
user	0m3.389s
sys	0m0.071s
```

After:
![image](https://github.com/user-attachments/assets/341b690e-35b5-47b0-ac1b-87c5cab094d4)
```
real	0m0.335s
user	0m0.284s
sys	0m0.052s
```

bocchino commented:
This is great! The performance gain is impressive, and it should help speed up the F Prime builds. Let's discussed the proposed changes to the spec.

Kronos3 commented:
I have pushed the following changes
- Clarified whitespace handling of multi-line string literals in the spec. The old behavior did not match the spec. We decided to update the spec to match the behavior.
- Updated lexer to allow trailing spaces after a line continuation, any other character will trigger an error
- Added a testcase for line continuation and multi-error reporting (updated `illegal-character` test in `fpp-syntax`)

Kronos3 commented:
More interesting benchmarks from our CI jobs:

Before
<img width="1364" alt="Screenshot 2025-04-29 at 11 34 00 AM" src="https://github.com/user-attachments/assets/73267a1d-4119-4452-93cf-4f4d1b949ab4" />

After
<img width="1364" alt="Screenshot 2025-04-29 at 11 33 52 AM" src="https://github.com/user-attachments/assets/39a1f004-b253-4805-b61d-77e6dac9cd81" />

Before
<img width="1346" alt="Screenshot 2025-04-29 at 12 47 04 PM" src="https://github.com/user-attachments/assets/d20c1e0f-4cc1-4a24-bb73-fb753e96ab0d" />

After
<img width="1346" alt="image" src="https://github.com/user-attachments/assets/1b4d1637-c301-4914-95fd-6f53b9dc4959" />

bocchino reviewed (approved):
Looks great! I made a few minor changes. When CI passes, I will merge it.

Kronos3 commented:
I neglected to put the most important benchmark here which is the cmake configuration time. I purged FppTest before each run and ran a native GraalVM build before/after:

Before:
```
-- Configuring done (23.9s)
-- Generating done (1.7s)
```

After:
```
-- Configuring done (10.4s)
-- Generating done (1.7s)
```
